### PR TITLE
モデルの軽量化（エラーあり）

### PR DESCRIPTION
**Pull request by Ayaka1156**

### CODE REVIEW

File: data_set_processing.ipynb [hunk 1]
```diff
@@ -7,38 +7,85 @@
...
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "Requirement already satisfied: fugashi in /usr/local/lib/python3.10/dist-packages (1.3.0)\n",
+            "Requirement already satisfied: ipadic in /usr/local/lib/python3.10/dist-packages (1.0.0)\n"
+          ]
+        }
+      ],
       "source": [
         "! pip install fugashi ipadic"
```
Comment (Last-Vega) [inline data_set_processing.ipynb:90]: この辺のpip istallとかは最上部で一つのセルにまとめるか，requirements.txtとかでまとめたいですね．
google colab環境にしない場合には，pyenvとかpoetryとかで管理できるとなお良いです

File: data_set_processing.ipynb [hunk 2]
```diff
@@ -295,29 +873,262 @@
...
+        }
+      ]
     },
     {
       "cell_type": "code",
       "source": [
         "import torch\n",
-        "from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer\n",
+        "\n",
         "from transformers import AdamW, get_linear_schedule_with_warmup\n",
```
Comment (Last-Vega) [inline data_set_processing.ipynb:1132]: こういうimport文とかも最上部のセルでまとめたいです

File: data_set_processing.ipynb [hunk 3]
```diff
@@ -431,25 +1226,53 @@
...
+        }
+      ],
       "source": [
         "# 分類モデルのため AutoModelForSequenceClassification を使用する\n",
         "# checkpoint と num_labels（クラス数） を指定する. 今回は、いずれも上で定義済み\n",
         "# - checkpoint = 'cl-tohoku/bert-base-japanese-whole-word-masking'\n",
         "# - num_labels = 8\n",
-        "config = AutoConfig.from_pretrained(checkpoint, num_labels=8)\n",
+        "# config = AutoConfig.from_pretrained(checkpoint, num_labels=8)\n",
         "model = AutoModelForSequenceClassification.from_pretrained(checkpoint)"
```
Comment (Last-Vega) [inline data_set_processing.ipynb:1251]: ここでmodelを再定義してしまっているので，せっかく上で軽量化のために6層で定義したモデルが上書きされちゃっている気がします．

File: data_set_processing.ipynb [hunk 4]
```diff
@@ -295,29 +873,262 @@
         "model = AutoModel.from_pretrained(checkpoint, config=config)"
```
Comment (Last-Vega) [inline data_set_processing.ipynb:873]: 既存のプロジェクトではAutoModelForSequenceClassificationが使われているようですが，AutoModelに変更したのって何か理由があるんでしたっけ？
軽量化のためみたいな理由であれば，読み込むモデル自体は変わっていないから劇的な改善は見られないと思うのでAutoModelForSequenceClassificationのままで良いのではないかなと思います．

File: data_set_processing.ipynb [hunk 5]
```diff
@@ -605,19 +1550,13 @@
...
-        "\n",
+        "            raise ValueError(\"Labels are missing or None in the inputs during training.\")\n",
         "        # モデルのフォワードパス\n",
         "        outputs = model(**inputs)\n",
-        "\n",
         "        # logitsから損失を計算\n",
-        "        loss = nn.CrossEntropyLoss()(outputs.logits, labels)\n",
-        "\n",
+        "        logits = outputs.logits\n",
+        "        loss = nn.CrossEntropyLoss()(logits, labels.view(-1))  # ラベルの形状を変更\n",
```
Comment (Last-Vega) [inline data_set_processing.ipynb:1559]: バッチサイズのエラーですが，`labels.view(-1)`でラベルの形状を1次元にしてしまっているのが原因だと思います．

たとえばバッチサイズ4でこの部分までの出力を確認すると，下の感じになります．


```python
class CustomTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        labels = inputs.get("labels", None)

        if labels is None:
            if "labels" not in inputs or inputs["labels"] is None:
                raise ValueError("Labels are missing or None in the inputs during training.")
            else:
                labels = inputs["labels"]

        # モデルのフォワードパス
        outputs = model(**inputs)
        # logitsから損失を計算
        # loss = nn.CrossEntropyLoss()(outputs.logits, labels)
        logits = outputs.logits
        print(logits)
        print(logits.shape)
        print(labels.shape)
        print(labels.view(-1).shape)
        loss = nn.CrossEntropyLoss()(logits, labels.view(-1))  # ラベル

        if return_outputs:
            return loss, outputs
        else:
            return loss
```
↓ 

```bash
tensor([[ 0.1013,  0.3497,  0.2316, -0.6505, -0.2727,  0.1579, -0.4748,  0.1752],
        [ 0.1655,  0.3399,  0.5123, -0.5844, -0.0353,  0.0723, -0.4650,  0.1123],
        [ 0.1872,  0.1636,  0.4679, -0.3973, -0.2780,  0.0323, -0.4288,  0.0695],
        [ 0.0474,  0.1745,  0.6009, -0.7317,  0.0181,  0.2515, -0.2737,  0.0817]],
       device='mps:0', grad_fn=<LinearBackward0>) # モデルが出力した，各文書（ミニバッチ4つ分の感情8クラス分の推定結果）
torch.Size([4, 8]) # tensorのサイズ
torch.Size([4, 8]) # labelのサイズ
torch.Size([32]) # label.view()した場合のサイズ
```


あとこれはもしかするとですが，南井さんの実装では
```python 
if return_outputs:
            return (loss, outputs)
        else:
            return loss.item() if hasattr(loss, "item") else loss
```
となっていますが，transformersの内部の処理でloss.itemを取り出す処理が走っていると思うので，条件分岐のelse以降はlossだけをreturnしちゃっていい気がします

**Comment by Last-Vega** [comment]
ローカル環境やgoogle colabを使ってもリソース不足や環境構築で困るということであれば，期間中だけクラウドのGPUマシンを借りてそれを使っちゃうのが良い気がします，

おすすめなのは，[runpod](https://www.runpod.io/),  [cerebrium](https://www.cerebrium.ai/), [modal](https://modal.com/)などです．少なくともrunpodは使わない時は起動を停止させておけばAWSとかよりは費用を格安で抑えられるので，colab上で全然実行できないみたいな状態が続くなら少し検討してみてください．（環境構築はできるに越したことはないですが，リソース不足みたいな問題は，今回のtech accelの本質とは少し違うので時間がもったいない気がします．）
